### PR TITLE
koordlet: unset cfs_quota for LSR/LSE pods

### LINKED ISSUE #489: [BUG] unexpected cfs throttles on cpuset pods if cfs_quota_us is set
**What happened**:

When pods are scheduled and started with static CPU binding policy and Linux cpuset controller, we expect that the pod/containers can use all the cpuset CPUs and not get throttled by Linux CFS Bandwidth Control. However, there are results which pods still have chance to get cfs throttled if `cpu.cfs_quota_us` is set (according to container's CPU limit).

e.g.

Pod A:

```yaml
      resources:
        limits:
          cpu: "1"
          memory: 2Gi
        requests:
          cpu: "1"
          memory: 2Gi
```

```bash
$ # print pod-level/container-level cfs_quota, cfs_period
$ cat /sys/fs/cgroup/cpu/kubepods.slice/kubepods-podxxx/cpu.cfs_quota_us
100000
$ cat /sys/fs/cgroup/cpu/kubepods.slice/kubepods-podxxx/cpu.cfs_period_us
100000
$ # print pod-level/container-level cpuset
$ cat /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-podxxx/cpuset.cpus
2
$ # run commands inside the pod
$ echo $$ > /sys/fs/cgroup/cpu/kubepods.slice/kubepods-podxxx/cgroup.procs
$ echo $$ > /sys/fs/cgroup/cpuset/kubepods.slice/kubepods-podxxx/cgroup.procs
$ while :; do :; done
$ # check cpu throttled count number (nr_throttled)
$ cat /sys/fs/cgroup/cpu/kubepods.slice/kubepods-podxxx/cpu.stat
nr_periods 3076
nr_throttled 34
throttled_time 12070000
```

**What you expected to happen**:

cpuset pods should not have cfs throttles. We may disable `cpu.cfs_quota_us` for the cpuset pods via the koordlet.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- App version: 0.6.x
- Kubernetes version (use `kubectl version`): 1.22
- Install details (e.g. helm install args):
- Others:
  - OS Version: Centos 7.9 (kernel 3.10) & Anolis OS 7 (kernel 4.19)

FYI:
- https://github.com/kubernetes/kubernetes/issues/67577
- https://github.com/torvalds/linux/commit/512ac999d2755d2b7109e996a76b6fb8b888631d

### PR DESCRIPTION
**Pull request by saintube**
Signed-off-by: saintube <saintube@foxmail.com>

### Ⅰ. Describe what this PR does

unset pod-level and container-level cgroup `cpu.cfs_quota_us` for cpuset pods (qos=LSR/LSE) to avoid unnecessary throttles

### Ⅱ. Does this pull request fix one issue?

fixes #489 

### V. Checklist

- [X] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [X] All checks passed in `make test`

### CODE REVIEW

File: pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go [hunk 1]
```diff
@@ -93,6 +98,48 @@ func (p *cpusetPlugin) SetContainerCPUSet(proto protocol.HooksProtocol) error {
...
+	podCtx := proto.(*protocol.PodContext)
+	if podCtx == nil {
+		return fmt.Errorf("pod protocol is nil for plugin %v", name)
+	}
+	req := podCtx.Request
+
+	// cpuset from pod annotation (LSE, LSR)
+	// NOTE: unset cfs quota for cpuset pods to avoid unexpected throttles.
+	// https://github.com/koordinator-sh/koordinator/issues/489
+	if cpusetVal, err := getCPUSetFromPod(req.Annotations); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go:111]: define an independent checking function in a new file for compatibility adaption.

File: pkg/util/pod.go [hunk 2]
```diff
@@ -276,3 +277,9 @@ func GetPodRequest(pod *corev1.Pod, resourceNames ...corev1.ResourceName) corev1
 func IsPodTerminated(pod *corev1.Pod) bool {
 	return pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed
 }
+
+// IsPodCPUBurstable checks if cpu burst is allowed for the pod.
+func IsPodCPUBurstable(pod *corev1.Pod) bool {
```
Comment (zwzhang0107) [inline pkg/util/pod.go:282]: put this into a independent file

File: pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go (hunk 1 above)
Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go]: done

File: pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go [hunk 3]
```diff
@@ -93,10 +98,44 @@ func (p *cpusetPlugin) SetContainerCPUSet(proto protocol.HooksProtocol) error {
...
+	podCtx := proto.(*protocol.PodContext)
+	if podCtx == nil {
+		return fmt.Errorf("pod protocol is nil for plugin %v", name)
+	}
+	req := podCtx.Request
+
+	// cpuset from pod annotation (LSE, LSR)
+	// NOTE: unset cfs quota for cpuset pods to avoid unexpected throttles.
+	// https://github.com/koordinator-sh/koordinator/issues/489
+	if cpusetVal, err := util.GetCPUSetFromPod(req.Annotations); err != nil {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go:111]: 这里用一个单独的方法来做判断比较合适

File: pkg/util/pod_qos_utils.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+package util
+
+import (
+	corev1 "k8s.io/api/core/v1"
+
+	apiext "github.com/koordinator-sh/koordinator/apis/extension"
+)
+
+func GetCPUSetFromPod(podAnnotations map[string]string) (string, error) {
```
Comment (zwzhang0107) [inline pkg/util/pod_qos_utils.go:25]: add comments for this file,
functions inside this file can be overwrite for extension

File: pkg/runtimeproxy/resexecutor/cri/utils.go [hunk 5]
```diff
@@ -74,7 +74,7 @@ func updateResource(a, b *v1alpha1.LinuxContainerResources) *v1alpha1.LinuxConta
 	if b.CpuPeriod > 0 {
 		a.CpuPeriod = b.CpuPeriod
 	}
-	if b.CpuQuota > 0 {
```
[the remark this answers is not in the document]
Comment (saintube) [reply pkg/runtimeproxy/resexecutor/cri/utils.go:77]: runtimeproxy should use non-zero value

File: pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go (hunk 3 above)
Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go]: done

File: pkg/util/pod_qos_utils.go (hunk 4 above)
Comment (saintube) [reply pkg/util/pod_qos_utils.go]: done

File: pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go [hunk 6]
```diff
@@ -93,10 +98,44 @@ func (p *cpusetPlugin) SetContainerCPUSet(proto protocol.HooksProtocol) error {
 	return nil
 }
 
-func getCPUSetFromPod(podAnnotations map[string]string) (string, error) {
-	podAlloc, err := ext.GetResourceStatus(podAnnotations)
-	if err != nil {
-		return "", err
+func (p *cpusetPlugin) SetPodCPUQuota(proto protocol.HooksProtocol) error {
```
Comment (zwzhang0107) [inline pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go:101]: naming: how about `UnsetPodCPUQuota`, `UnsetContainerCPUQuota`.
does not need to be member method

Comment (saintube) [reply pkg/koordlet/runtimehooks/hooks/cpuset/cpuset.go]: done

**Pull request merged**